tests/integration/test_bundle.py

```diff
@@ -0,0 +1,75 @@
+import logging
+from pathlib import Path, PosixPath
+import shlex
+import yaml
+
+from pytest_operator.plugin import OpsTest
+from helpers.localize_bundle import localize_bundle_application
+
+BUNDLE_CHARMS = [
+    "kfp-api",
+    "kfp-persistence",
+    "kfp-profile-controller",
+    "kfp-schedwf",
+    "kfp-ui",
+    "kfp-viewer",
+    "kfp-viz",
+]
+
+CHARM_PATH_TEMPLATE = "{basedir}/charms/{charm}/"
+
+log = logging.getLogger(__name__)
+
+
+async def test_build_and_deploy(ops_test: OpsTest, request):
+    if ops_test.model_name != "kubeflow":
+        raise ValueError("kfp must be deployed to namespace kubeflow")
+
+    # Get/load template bundle from command line args
+    bundlefile = Path(request.config.getoption("bundle"))
+    basedir = Path("./").absolute()
+    bundle = yaml.safe_load(bundlefile.read_text())
+
+    # Build the charms we need to build
+    charms_to_build = {
+        charm: Path(CHARM_PATH_TEMPLATE.format(basedir=str(basedir), charm=charm))
+        for charm in BUNDLE_CHARMS
+    }
+    log.info(f"Building charms for: {charms_to_build}")
+    built_charms = await ops_test.build_charms(*charms_to_build.values())
+    log.info(f"Built charms: {built_charms}")
+
+    # Edit the bundle, splicing our locally built charms into it
+    for charm, path in built_charms.items():
+        bundle = localize_bundle_application(bundle, application=charm, charm_file=path)
+
+    # Deploy the bundle
+    # TODO This should be easier... but it isn't.  Can either use the `juju bundle` integration, or
+    #      shell out to juju (like k8s charms do).  For now, I'm doing the latter
+    #      Observability has a function in their CI that is similar.  Use that?
+    bundle_dst_file = write_bundle_file(bundle, ops_test.tmp_path / "bundles" / "bundle.yaml")
+
+    model = ops_test.model_full_name
+    # TODO: Figure out when we need --trust rather than blanket-applying it
+    cmd = f"juju deploy -m {model} --trust {bundle_dst_file}"
+    log.info(f"Deploying bundle to {model} using cmd '{cmd}'")
+    rc, stdout, stderr = await ops_test.run(*shlex.split(cmd))
```

Comment:
All of this seems like something we should add to pytest-operator and standardize somehow.

Reply:
100% agree.  pytest-operator has some similar features, but they're a bit different.  They don't do this "convert a publishable bundle into partly-local", so I think this would help.  I'm thinking we gain a little experience then push the final product to pytest-operator?